File: 726. Number of Atoms.py
import collections

class Solution(object):
    def countOfAtoms(self, formula):
        stack = [collections.defaultdict(int)]
        i = 0
        n = len(formula)
        
        while i < n:
            if formula[i] == '(':
                stack.append(collections.defaultdict(int))
                i += 1
            elif formula[i] == ')':
                i += 1
                start = i
                while i < n and formula[i].isdigit():
                    i += 1
                multiplicity = int(formula[start:i] or 1)
                top = stack.pop()
                for element in top:
                    stack[-1][element] += top[element] * multiplicity
            else:
                start = i
                i += 1
                while i < n and formula[i].islower():
                    i += 1
                name = formula[start:i]
                start = i
                while i < n and formula[i].isdigit():
                    i += 1
                multiplicity = int(formula[start:i] or 1)
                stack[-1][name] += multiplicity
        
        result = []
        for name in sorted(stack[-1]):
            result.append(name)
            count = stack[-1][name]
            if count > 1:
                result.append(str(count))
        
        return ''.join(result)

File: 726. Number of Atoms.cpp
class Solution
{
public:
    string countOfAtoms(string formula)
    {
        map<string, int> counts = parseFormula(formula);
        string result;
        for (auto &[elem, cnt] : counts)
        {
            result += elem;
            if (cnt > 1)
                result += to_string(cnt);
        }
        return result;
    }

private:
    map<string, int> parseFormula(const string &formula)
    {
        stack<map<string, int>> stack;
        stack.push(map<string, int>());
        int n = formula.size();

        for (int i = 0; i < n;)
        {
            if (formula[i] == '(')
            {
                stack.push(map<string, int>());
                i++;
            }
            else if (formula[i] == ')')
            {
                map<string, int> top = stack.top();
                stack.pop();
                i++;
                int i_start = i;
                while (i < n && isdigit(formula[i]))
                    i++;
                int multiplicator = i > i_start ? stoi(formula.substr(i_start, i - i_start)) : 1;
                for (auto &[elem, cnt] : top)
                {
                    stack.top()[elem] += cnt * multiplicator;
                }
            }
            else
            {
                int i_start = i;
                i++;
                while (i < n && islower(formula[i]))
                    i++;
                string elem = formula.substr(i_start, i - i_start);
                i_start = i;
                while (i < n && isdigit(formula[i]))
                    i++;
                int cnt = i > i_start ? stoi(formula.substr(i_start, i - i_start)) : 1;
                stack.top()[elem] += cnt;
            }
        }

        return stack.top();
    }
};
